### ISSUE TITLE
[Idea] Zero-shot object detection models?

### ISSUE THREAD

**Issue by afaulconbridge** [blakeblackshear/frigate#10975]
**Describe what you are trying to accomplish and why in non technical terms**
I want to be able to use [zero-shot object detection](https://huggingface.co/models?pipeline_tag=zero-shot-object-detection) / [zero-shot image classification](https://huggingface.co/models?pipeline_tag=zero-shot-object-detection) so that I can record wildlife. IR footage is poorly supported and many training data sets are poorly labelled for uncommon animals (fox or hedgehog for example).

**Describe the solution you'd like**
I'm curious if anyone has considered this before, and if so what pitfalls there might be. Or if there are any plans for "plug ins" or similar ways to extend Frigate. I'll have a go at my own fork with an object-detection implementation, and I might be able to get something working locally for my use case, but I probably won't be able to generalise it without help.

**Describe alternatives you've considered**
I could train my own model (Frigate+ style) but I'd rather not as its a lot of work to prepare, update, etc. Zero-shot approaches seem to be improving faster, and off-the-shelf models improving too.

**Additional context**
I've been building my own system from scratch, but then I discovered Frigate and it already does so much of the "boilerplate" much better than I could. Nice job!

**Comment by NickM-27** [comment]
Frigate already works with deepstack (CodeProject.AI) so that would work for many cases, Frigate also supports using any model you like as long as it is compatible with the detector of your choosing. I am not sure what exactly the request is here

**Comment by afaulconbridge** [comment]
Deepstack, and the other models in Frigate as far as I can see, work with _pre-defined and pre-trained_ [labels](https://docs.frigate.video/configuration/objects). So it can never detect a hedgehog, because it has no concept of what a hedgehog could be, because there was no hedgehog in the training labels.

Zero-shot models are only given the categories at execution time, and can work with things that they've never seen before - similar to large-language models. Broadly, a zero-shot model knows that a hedgehog is a small spiky spherical mammal, so when it sees a picture of a hedgehog it can give the right answer - even if the training data didn't contain any hedgehog pictures and only included picture of things that were either small or spiky or spherical or mammals.

I'm going to mess around with using zero-shot models in Frigate. There's no IRC, discord, reddit, email, or other contact/community details I could see - so a GitHub issue it is. I know Frigate is MIT licensed, so technically I don't need to tell you, but I thought it'd be nice to say something (and might save me a bunch of time if it was trivially impossible for some reason). Plus if there was anyone else who has/is/will try this it might be useful to cooperate.

**Comment by fgoodwin** [comment]
https://www.reddit.com/r/frigate_nvr/

Additional tags are coming Eventually™ but I also am interested in more generalized detection for wildlife, so if you get it working definitely write something up.

**Comment by teajaysee** [comment]
did you get anywhere with this as im interested in similar?

**Comment by afaulconbridge** [comment]
TL;DR no @teajaysee , its a mess. 😢 

I've been playing around with this for almost two years now, which in AI years is about 14. Nothing has quite performed well enough to be deployable. For example, hedgehogs are almost never detected by any off-the-shelf model unless it is so eager that twigs and tufts of grass are reported as hedgehogs. On the other hand, cats are easy to detect - even when its a fox or a badger. Becuase models are trained to detect things, and on a video feed not much is happening most of the time, they tend to throw a lot of false positives too.

Some models run fast, but are very wrong (e.g. YOLO-E - which is also a pretty poorly written package so is a nightmare to work with for any non-trivial scenario)
Some models generate a vector from the entire frame and compare to a text vector (e.g. CLIP variants, BioCLIP in particular). However, they work only at a fixed low resolution so unless its something right in front of the camera, its missed. Matching frames to stored images rather than text vectors works better, but still gets it wrong a lot of the time seemingly at random. This approach also has potential to record first and then go back and look through a collection of clips retrospectively later, which is an interesting alternative.
Some models are very compute intensive, and don't quite fit what I need (e.g. SAM).
Some are a student project that is neither complete nor packaged nor documented.
All of these come with a range of formats & libraries, and trying to use them with other new tech also creates a mess because nothing is stable.

The latest I've been experimenting with is Moondream. This is more like a (small) LLM with high-resolution vision capabilities and it is accurate for finding bounding boxes, but doesn't have confidence scores (so tracking and multiple classes need special handling). It is slow (1-3fps on 4070ti) and scales poorly as it has to reprocess each image+class combination from scratch.

I have yet to find any model that can handle video natively. By that I mean using the temporal nature of an image sequence, rather than looking at a frame in isolation. Seems to me as if reusuing some kind of attention vector from one frame to another would be possible, and there should be plenty of accessibilty based video description material for training. 

I've also fallen down related rabbit holes along the way, such as camea specs; resolution vs frame rate - and how ISO and sensor size and appature size interact with that, colour star light vs B&W IR, motion sensor or not, DIY vs off-the-shelf, and cost tradeoffs with all of the above.
